src/tiledb/cloud/geospatial/ingestion.py

```diff
@@ -0,0 +1,1287 @@
+                logger.debug("finding metadata for %r", pth)
+                with fiona.open(pth) as src:
+                    if crs is None:
+                        crs = src.crs
+                    else:
+                        if src.crs != crs:
+                            raise ValueError("All datasets need to be in the same CRS")
+
+                    if schema is None:
+                        schema = src.schema
+                    else:
+                        if src.schema != schema:
+                            raise ValueError(
+                                "All datasets need to have the same schema"
+                            )
+
+                    # 2.5 is supported internally but not needed for ingest
+                    extents = BoundingBox(
+                        minx=src.bounds[0],
+                        miny=src.bounds[1],
+                        maxx=src.bounds[2],
+                        maxy=src.bounds[3],
+                    )
+                if full_extents:
+                    full_extents = BoundingBox(
+                        *shapely.unary_union(
+                            [
+                                shapely.box(*full_extents.bounds),
+                                shapely.box(*extents.bounds),
+                            ]
+                        ).bounds
+                    )
+                else:
+                    full_extents = extents
+
+            return GeoMetadata(
+                extents=full_extents, crs=crs, paths=sources, geometry_schema=schema
+            )
+
+
+def get_raster_metadata(
+    sources: Iterable[os.PathLike],
+    *,
+    config: Mapping[str, Any] = None,
+    dst_tile_size: [int] = RASTER_TILE_SIZE,
+    res: Tuple[float, float] = None,
+    verbose: bool = False,
+    id: str = "raster_metadata",
+    trace: bool = False,
+    log_uri: Optional[str] = None,
+) -> GeoMetadata:
+    """Return geospatial metadata for a sequence of input raster data files
+
+    :param sources: iterator, paths or path to process
+    :param config: dict, configuration to pass on tiledb.VFS
+    :param dst_tile_size: int, tile size of the destination mosaic
+    :param verbose: bool, enable verbose logging, default is False
+    :param trace: bool, enable trace logging, default is False
+    :param log_uri: Optional[str] = None,
+    :Return: GeoMetadata, a populated GeoMetadata objects with blocks
+             and the files that contribute to each block
+    """
+    if len(sources) == 0:
+        raise ValueError("Input raster datasets required")
+
+    logger = get_logger_wrapper(verbose)
+    full_extents = None
+    crs = None
+    num_bands = None
+    do_res_check = True
+    dtype = None
+
+    if res:
+        xres, yres = res
+        do_res_check = False
+    else:
+        xres = None
+        yres = None
+
+    def group_by_raster_block(
+        meta: Sequence[GeoMetadata],
+    ) -> Tuple[GeoBlockMetadata, ...]:
+        nonlocal logger
```

Comment:
Since this doesn’t change the value of `logger` (i.e., it doesn’t assign to it/re-bind it), this statement is not needed; the closure captures `logger`.

Reply:
👍